### PR TITLE
Change TS shard allocation daily

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
==COMMIT_MSG==
Targeted Sweep shard allocation for writes now changes every day. This should alleviate issues caused by skewed write patterns where too many writes consistently go to the same shard.
==COMMIT_MSG==

**Implementation Description (bullets)**:
Include the hash of the current date in the shard calculation.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Existing tests cover the basics. The shard allocation method is only called in one place in production code, so it is straightforward to verify correctness.

**Concerns (what feedback would you like?)**:
When this is initially rolled out, and at midnight every day this may cause some `WriteInfo`s for the same cell to be allocated to two separate shards. Similarly, significant clock skew could cause the same issue. This could reduce the batching effectiveness since whenever this happens we would have to use two range tombstones (one for each shard) for the cell where earlier one may have been enough. This is an insignificant perf regression, and easily outweighed by better load balancing that we get.

This could make some of the hackier tests fail if run at the end of day -- we can address those if and when they happen, doing it proactively is too much effort.

**Where should we start reviewing?**:
tiny

**Priority (whenever / two weeks / yesterday)**:
Not super urgent, but it would be nice to roll out this week.

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/WriteInfo.java [hunk 1]
```diff
@@ -49,7 +51,18 @@ default TimestampRangeDelete toDelete(Sweeper sweeper) {
...
+    }
+
+    /**
+     * The purpose of the rotating hash calculation is to redistribute shards every day to alleviate issues caused by
+     * imbalanced write patterns overloading few shards.
+     */
+    default int dayRotatingHash() {
+        int hash = 5381;
+        hash = hash * 31 + writeRef().cellReference().goodHash();
+        hash = hash * 31 + LocalDate.now(ZoneId.of("UTC")).hashCode();
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/WriteInfo.java:64]: Discussed offline: this hash isn't good when you have a multiple of 31 shards (which is very common)!

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/CellReference.java [hunk 2]
```diff
@@ -35,9 +35,9 @@ public abstract class CellReference {
      */
     public int goodHash() {
         int hash = 5381;
-        hash = hash * 31 + tableRef().hashCode();
-        hash = hash * 31 + rowHash.hashBytes(cell().getRowName()).asInt();
-        hash = hash * 31 + colHash.hashBytes(cell().getColumnName()).asInt();
+        hash = hash * 1439 + tableRef().hashCode();
```
Comment (jeremyk-91) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/CellReference.java:38]: fair enough: `1439` is more shards than allowed

**Review by jeremyk-91 — COMMENTED** [review]
Should be fine: but for confirmation - this change is allowed even though it makes the hashing function unstable, because of range tombstones it is not a matter of correctness that the entries for a cell may be distributed across rows, correct?

**Review by jeremyk-91 — APPROVED** [review]
As discussed: This should be fine, we've checked the perf changes it might bring about. There is a technicality here where the tests might break if run at midnight 🔥 but not going to block on that.

**Pull request merged**